### PR TITLE
Simplify Interfaces Into Smaller, Useful Ones

### LINKED ISSUE #163: Create Common Testing Interface that all Sharding Actors Will be Able to Use
Currently the mock `smcClient` is used for testing both in the notary and proposer packages. 
```go
type smcClient struct {
	smc         *contracts.SMC
	depositFlag bool
	t           *testing.T
       backend   *simulated.backend
}

```
Instead of repeating it for testing in both notary and proposer, it would be better to define it in a common package and then use it from there.

So the mock smcClient , and its methods can be defined in `sharding/client/testing`  and have its methods defined something like this:
```go
func (s *smcClient) Account() *accounts.Account {
	return &accounts.Account{Address: addr}
}

func (s *smcClient) SMCCaller() *contracts.SMCCaller {
	return &s.smc.SMCCaller
}

func (s *smcClient) ChainReader() ethereum.ChainReader {
	return nil
}

func (s *smcClient) Context() *cli.Context {
	return nil
}
```
This will be blocked by #131 and #111 as it can only be done when they are completed

### LINKED ISSUE #164: Implement ChainReader in Tests
Currently in our testing we do not implement the chainreader interface:
```go
type ChainReader interface {
	BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error)
	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
	HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error)
	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
	TransactionCount(ctx context.Context, blockHash common.Hash) (uint, error)
	TransactionInBlock(ctx context.Context, blockHash common.Hash, index uint) (*types.Transaction, error)

	// This method subscribes to notifications about changes of the head block of
	// the canonical chain.
	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (Subscription, error)
}
```
This will be needed to write and implement more expressive and complete tests. As we will need to retrieve blocks or headers using either its number or hash, and the current mock `backend` does not provide that ability. 

One possible solution would be to use the reflect package to use the unexported fields in the mock backend 
```go
val := reflect.value(backend)
blkchn := val.FieldByName("blockchain").Elem()
```

### LINKED ISSUE #183: Breakdown Large Interfaces Into Smaller, More Useful Ones
Hi all,

Part of effective go is the usage of interfaces as a way to enable easy refactoring and reusability of code through layers of abstraction. Interfaces are ideal when mocking certain methods in order to unit test small/medium size functions, yet we often end up having very large interfaces that map almost 1-to-1 to their implementation. That is, we define items such as:

```go
type Client interface {
    Account() *accounts.Account
    CreateTXOpts(value *big.Int) (*bind.TransactOpts, error)
    SMCCaller() *contracts.SMCCaller
    SMCTransactor() *contracts.SMCTransactor
    SMCFilterer() *contracts.SMCFilterer
    TransactionReceipt(common.Hash) (*types.Receipt, error)
    ChainReader() ethereum.ChainReader
    DepositFlag() bool
    SetDepositFlag(deposit bool)
    DataDirPath() string
    Sign(hash common.Hash) ([]byte, error)
    GetShardCount() (int64, error)
}
```
_in sharding/mainchain/smc_client.go_

Which we then use in functions such as:
```go
func isAccountInNotaryPool(client mainchain.Client) (bool, error) {
	account := client.Account()
	// Checks if our deposit has gone through according to the SMC.
	nreg, err := client.SMCCaller().NotaryRegistry(&bind.CallOpts{}, account.Address)
	if !nreg.Deposited && err != nil {
		log.Warn(fmt.Sprintf("Account %s not in notary pool.", account.Address.String()))
	}
	return nreg.Deposited, err
}
```
_in sharding/notary/notary.go_

Notice how we have an entire `Client` interface, yet this function (among others), only use the `SMCClient()` property and the `Account()` property. When testing this method, we have to mock an entire struct and include every single method in the `Client` interface when we could instead be more effective.

Instead, we could define a 
```go
type ContractCaller interface {
    SMCCaller()
}
```

as well as a 
```go
type AccountFetcher interface {
    Account()
}
```

and refactor this function to take in these interfaces. Testing in this case becomes incredibly easy, and the `*mainchain.SMCClient` struct ends up being an implementation of a composition of these smaller interfaces. I already started refactoring some aspects of this in my PR #172 as it was nearly impossible to test certain goroutines without having to mock every single method across a variety of interfaces.

Let's keep this in mind moving forward.

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This is a PR referencing #183 which aims to use effective go practices of using small, but useful interfaces in our code.

### What Was Wrong?

We had giant monolith interfaces such `Client` that mapped 1-to-1 to the structs implementing them. When writing tests, we had to mock a lot of unnecessary methods and were no better off than using the struct itself.

### What Was Fixed?

We make smaller interfaces that only require specific, useful mocks for the tests in which they are used. Ex:

```go
// ContractCaller defines an interface that can read from a contract on the
// Ethereum mainchain as well as call its read-only methods and functions.
type ContractCaller interface {
	SMCCaller() *contracts.SMCCaller
	GetShardCount() (int64, error)
	Account() *accounts.Account
}

// ContractTransactor defines an interface that can transact with a contract on the
// Ethereum mainchain as well as call its methods and functions.
type ContractTransactor interface {
	SMCTransactor() *contracts.SMCTransactor
	CreateTXOpts(value *big.Int) (*bind.TransactOpts, error)
}

// Reader defines an interface for a struct that can read mainchain information
// such as blocks, transactions, receipts, and more. Useful for testing.
type Reader interface {
	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
}
```

This resolves #183, resolves #163, and resolves #164.

### CODE REVIEW

File: sharding/mainchain/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	ContractCaller
+	ContractTransactor
+}
+
+// ContractCaller defines an interface that can read from a contract on the
+// Ethereum mainchain as well as call its read-only methods and functions.
+type ContractCaller interface {
+	SMCCaller() *contracts.SMCCaller
+	GetShardCount() (int64, error)
+	Account() *accounts.Account
```
Comment (terenc3t) [inline sharding/mainchain/interfaces.go]: We don't really Account address from contract. At the same time, I'm not sure where else to put it. Do you think the Reader interface is more appropriate for `Account()`?

File: sharding/notary/notary.go [hunk 2]
```diff
@@ -14,69 +12,34 @@ import (
...
-		}
-	}
-}
-
 // checkSMCForNotary checks if we are an eligible notary for
 // collation for the available shards in the SMC. The function calls
 // getEligibleNotary from the SMC and notary a collation if
 // conditions are met.
-func checkSMCForNotary(client mainchain.Client, head *types.Header) error {
+func checkSMCForNotary(caller mainchain.ContractCaller, head *types.Header) error {
```
Comment (prestonvanloon) [inline sharding/notary/notary.go]: Why not pass in SMCCaller and the account as a seperate argument (since it doesnt seem to be related to the contract at all)?

File: sharding/notary/notary.go [hunk 3]
```diff
@@ -14,69 +12,34 @@ import (
 	shardparams "github.com/ethereum/go-ethereum/sharding/params"
 )
 
-// SubscribeBlockHeaders checks incoming block headers and determines if
```
Comment (prestonvanloon) [inline sharding/notary/notary.go]: Why did you delete this method?

Comment (prestonvanloon) [reply sharding/notary/notary.go]: Why did you delete this comment?
